### ISSUE TITLE
Vectorisation

### ISSUE THREAD

**Issue by TomMaullin** [cvxopt/cvxopt#146]
I have been using `CVXOPT` lately and have found it brilliant for individual sparse cholesky decompositions. However, I was just wondering if `CVXOPT` has any plans in vectorising/broadcasting operations across multiple matrices? i.e. much like [numpy's broadcasting rules](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html). 

As I mentioned my work currently involves the sparse cholesky factorization and I am losing a large amount of time as I am having to loop over many different arrays (all of the same dimension, and even sparsity structure in some circumstances) to perform the sparse cholesky on each matrix individually. If there was a way these decomposition's could be vectorized to avoid python loops it would be greatly appreciated.

**Comment by martinandersen** [comment]
I am not sure I understand which operation(s) you would like to support broadcasting. Do you need to factorize multiple matrices with the same sparsity pattern?

**Comment by TomMaullin** [comment]
Yes, in my case they have the same sparsity pattern and therefore the same permutation vector P. Ideally what would be broadcastable is the cholesky decomposition on sparse matrices itself. I.e. Suppose I had a 3D matrix S; I wish to compute for every `i`:

```
cvxopt.cholmod.solve(cvxopt.cholmod.numeric(S[i,:,:], cvxopt.cholmod.symbolic(S[i,:,:],perm=P)),S[i,:,:], sys=k)
```

Currently the only way I can do this is by creating a 3D sparse matrix in the `sparse` python package and converting every individual 2d slice of S to a `cvxopt` sparse matrix and then performing the sparse cholesky individually on each matrix. All the individual operations are very fast but my code is extremely slow due to the looping and conversion between sparse datatypes.

**Comment by lciti** [comment]
I know it's an old post but in case someone has the same issue.

- If the sparsity pattern of all matrices is the same, a more efficient way is to compute one symbolic factorisation at the very beginning (not for each matrix in the stack and not for each loop, just once).
- Avoid converting between 3d non-cvxopt sparse matrices (e.g. from scipy.sparse or others) and cvxopt sparse matrices, but rather use a python list of cvxopt sparse matrices.

```python
# once, before calling any cvxopt solver
S_chol = cvxopt.cholmod.symbolic(S[0], perm=P)
...
# called at each iteration, e.g. because inside kktsolve()
for i in range(k):
     cvxopt.cholmod.numeric(S[i], S_chol)
     cvxopt.cholmod.solve(S_chol, b[i], sys=0)
```
